Text based_Random+Greedy+KMeans+Exact

zohrehraziei opened the pull request:
Hi,

The current PR contains the random optimization method with the following options (first row from Table 1 and 2):
rand(t=max_num_features)​ (no difference with unicost and diverse)
rand(t=unicost())
rand(t=diverse())

The code uses a random optimization method to select the most relevant features from a text corpus. The algorithm randomly selects a set of columns, calculates the number of rows covered by these columns, and selects the set that covers the most rows. The user can set the selection size or leave it as None for the algorithm to calculate automatically (by solving a set cover problem with unicost or diversity cost). The diversity cost of each feature calculates using K-Means clustering. 

Thank you!
Zohreh

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+        # Defualt values
+        num_features: Union[int, None]
+        featurization_method: TextWiser = TextWiser(Embedding.TfIdf(min_df=10), Transformation.NMF(n_components=30))
+        optimization_method: str = "random"
```
let's make the default as "exact" + "diverse" --which corresponds to the paper

zohrehraziei replied:
Then we need to add cost_metric=None for tests without a cost metric requirement.

skadio replied:
Not sure how the default behavior is related to None cost metric. 

Let's please not add a None to cost. Instead; 

For methods that don't depend on cost (say random()), we will document that clearly in the pydoc. 
Then, in the tests, we will write a test with each cost and verify that the results are the same (since the method is not dependant on cost)

zohrehraziei replied:
The default value of "None" for the cost metric has been removed. Additionally, an "if" condition has been added to the "_process_df" function to handle a scenario where the optimization method is "random", and we have a cost metric (either "unicost" or "diverse"), and the user has specified the number of features. In this case, then the text featurization method will NOT be executed.

To ensure that the results are consistent, a test has been created using two instances of "TextBased" - one with "cost_metric" set to "unicost" and the other with "cost_metric" set to "diverse". Both instances have the same number of features specified. The test checks whether the resulting indices are the same using "np.allclose()". If the indices are the same, the test passes.

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
+                           The number of selected times (num_features = t):
+                                - t = maximum number of features defines by user.
```
defineD

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+                - Greedy optimization with unicost as cost_metric
+
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
```
"items" is a terminology from the paper, but Selective users don't know that. so maybe, say set of items/features

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
+                           The number of selected times (num_features = t):
```
TYPO items (or better features)

skadio replied:
No need to intent this further.. Keep everyhing at the same level with "This method finds.." Otherwise it is getting too nested to read

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+        References:
+        [1] Kadioglu et. al., Optimized Item Selection to Boost Exploration for Recommender Systems, CPAIOR'21
+
+        The parameters corresponding to the methods in the paper [1]
```
I think we can remove this block from 294 to 304.. We will set the paper parameters as the default values.

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+                                - t = maximum number of features defines by user.
+                                      The cost_metric input argument needs to ignore in this case.
+                                - t = None: the number of feature computed by solving a set cover problem with
+                                      cost metrics (unicost or diverse)
```
with the "given" cost metric.

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
+                           The number of selected times (num_features = t):
+                                - t = maximum number of features defines by user.
```
this is not the "maximum" is it? When t is given, we return exactly t columns, no?

zohrehraziei replied:
Yes, this is not Max. Anywhere that the number of features is an input from the user, it should be num_features.

skadio commented on README.md:
```diff
@@ -40,13 +41,14 @@ print("Scores:", list(selector.get_absolute_scores()))
+|    [Statistical Analysis](https://scikit-learn.org/stable/modules/feature_selection.html#univariate-feature-selection)     |                                                                                                             [ANOVA F-test Classification](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html) <br> [F-value Regression](https://scikit-l ...
+|                             [Linear Methods](https://en.wikipedia.org/wiki/Linear_regression)                              |                                                                                                   [Linear Regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html?highlight=linear%20regression#sklearn.linear_model.LinearRegr ...
+|                          [Tree-based Methods](https://scikit-learn.org/stable/modules/tree.html)                           | [Decision Tree](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn.tree.DecisionTreeClassifier) <br> [Random Forest](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=ra ...
+|  [Text-based Methods](https://link.springer.com/chapter/10.1007/978-3-030-78230-6_27)  |                                                                                                                                                                                                                                                                                                                      ...
```
Let's call it "TextWiser featurization_method" so that the parameter name is used as is, like the other two params

zohrehraziei replied:
The name in README.md looks good to me! 
Should I need change the featurization_method in the code by TextWiser featurization_method"?

skadio replied:
Not in the code.. I meant on line 51 here in the README. 

The way it is written now (with the link) does not tell the user what's the name of the parameter they can use. 

You can replace the first parameter with "TextWiser (link this) `featurization_method`" so that a user looking at the readme sees that the 3 parameters are called featurization_method, optimization_method, and cost_metric 

OR .. even better, to be consistent with other lines you can write this the other way around.. " `featurization_method` = TextWiser(and link this)

skadio commented on README.md:
```diff
@@ -30,6 +30,7 @@ selector = Selective(SelectionMethod.Correlation(threshold=0.5, method="pearson"
+selector = Selective(SelectionMethod.TextBased(num_features=3))
 
 # Feature reduction
 subset = selector.fit_transform(data, label)
```
now.. this example with this data/labal is not gonna work with the TextBased selector, no?

skadio replied:
I think we can create a separate section after Benchmarking to show an example of TextBased selection and make it explicit what data and labels look like in the input with a small toy problem

zohrehraziei replied:
Yes, it does not work. Great Idea! will add a section after Benchmarking and will let you to review.

zohrehraziei replied:
The section added. Please let me know with your feedback.

skadio replied:
then you want to delete the above line, no?

zohrehraziei replied:
I added a reference for the sub-section text-based selective in Selective: Feature Selection Library. Since people first take a look at Quick Start and may lose where the one for text-based is added.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+
+from textwiser import TextWiser, Embedding, Transformation
+
+
```
dangling lines

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+
+    @staticmethod
+    def _verify_input(cost, matrix):
+        assert (cost is not None), "Data Error: cost cannot be none"
```
shall we replace asserts with check_true() function?

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+class ContentSelector:
+    """ Content Selector.
+
+    This class contains a content selection algorithm (max_cover) that maximizes the diversity among the selected
```
I think this paragraphs need a bit of adjusting. This class does not contain a single max_cover algorithm, does it?

zohrehraziei replied:
yes, modified.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+
+    def run_content_selection(self,
+                              input_df: pd.DataFrame,
+                              categories: List[int], featurization_method: TextWiser, method: str = "random",
```
move featurization_method to next line

skadio replied:
shall we replace method with "optimization_method" so that it is consistent with the outer TextBased class? Also, move it to other line

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+            - "exact" maximizes content diversity and minimizes the number of content to cover all content labels.
+            - "kmeans" clusters the content into selection_size number of clusters and then selects the items closest
+              to the centroid of each of the clusters. This method does not consider the content categories/labels.
+            - "greedy" performs greedy heuristic selecting items with max unit coverage until all items are covered
```
The kmeans line has the "This method does not consider the content labels .." and analogously, greedy can have "This method does not consider the diversity"

zohrehraziei replied:
Shall we remove the definitions of optimization algorithms from here? We extensively explain each of them in the class TextBased in selector.py.

skadio replied:
YES! I like that suggestion. There are good information here too, so please "merge" all them under selector.py

skadio replied:
(and here drop a comment that "see TextBased in Selector.py" or sth

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        -------
+        No return
+       --------
+        * categories is the list of labels. code for labels with categories moved to utils.py
```
not following this line with the asterix *

zohrehraziei replied:
I modified the text to: "We have implemented two versions of this function. In the version in the text_based.py, the categories parameter actually refers to the labels (rather than categories) to be covered. The version of code for generating labels from categories has been moved to utils.py."

In the _process_df function in utils.py, each label is created by concatenating multiple category values from the input dataframe, which means each label can have different categories. The categories for each label are defined by the categories parameter, which is a list of column names in the input dataframe that contain the category values.

 The _process_df function is currently used, we assume that categories is a numpy array of labels, and not a list of columns in the input data frame containing the labels. Therefore, if we assume that we already have a matrix of labels and there are no categories.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+    def _process_df(self, input_df: pd.DataFrame, categories: List[int], method: str, cost_metric: Optional[str]) \
+            -> NoReturn:
+        """
+
```
can we add a line at the top, here, to say more on what this method does..

skadio commented on feature/utils.py:
```diff
@@ -78,6 +78,36 @@ def normalize_columns(df: pd.DataFrame):
     return df_normalized.fillna(0)
 
 
+def _process_df(input_df: pd.DataFrame, categories: List[str], feature_column: str, selection_size: int) \
```
what's the difference between this process_df function and the other process_df function?

skadio replied:
do we need both?

zohrehraziei replied:
No, we do not need both. I explained in a few lines at the tope of _process_df function in text_based.py

skadio replied:
remind me again what's the difference between this process and the other one text_based..

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        selected_features = [col for i, col in enumerate(text_columns) if i in selected_indicies]
+        self.selected_features_ = selected_features
+
+
```
extra line

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        self.cost_metric = cost_metric
+        self.content_selector = ContentSelector(selection_size=num_features, seed=123456, trials=100, verbose=True)
+
+    def fit(self, data: pd.DataFrame, labels: Union[pd.Series, pd.DataFrame]) -> NoReturn:
```
can the labels be Series?

zohrehraziei replied:
Removed! As we need rows with different labels, we need to use a Pandas DataFrame, and a Pandas Series is a one-dimensional labeled array.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        self.featurization_method = featurization_method
+        self.optimization_method = optimization_method
+        self.cost_metric = cost_metric
+        self.content_selector = ContentSelector(selection_size=num_features, seed=123456, trials=100, verbose=True)
```
you are missing to use the self.seed here

skadio replied:
after all, seed is essentially the most important part of the Random selector :D

zohrehraziei replied:
Fixed! Thank you for mentioning it!

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        if self.selection_size is not None:
+            assert (self.selection_size <= self._num_cols), "Process Data Error: selection_size cannot exceed num columns"
+
+
```
extra lines..

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+
+        return best_selected
+
+
```
extra line

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        # Return
+        return selected
+
+
```
extra line

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        self.trials = trials
+        self.verbose = verbose == 1
+
+
```
extra line

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        best_selected = []
+        best_covered = 0
+
+        if self.selection_size is None:
```
You can create a private method for the inner loop, as we will keep doing this operation for many other places where selection_size is None.. 
So the code will simplift to: 
```
selected_size = _calculate_selection_size() if selection_size is None else selection_size
```

skadio replied:
There is a more advanced version that we can do, but I am not sold on it (because it splits the logic)

you can move that one-liner to the constructor, so that whenver a None selection_size comes in, as the input, we do that operation once in the constructor and not worry about it here.. EXCEPT when the option is "exact"... 

Let's not go there for now, but I am just making the connection

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        num_row_covered = self._get_num_row_covered(best_selected)
+        coverage = num_row_covered / self._num_rows
+
+        if self.verbose:
```
I assume the default verbosity will be set to False, so that the user don't see some random printouts..

zohrehraziei replied:
The 'verbose' parameter in the '__init__' method of class ContentSelector is set to 0 by default.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+
+    """
+
+    def __init__(self,  selection_size: int, seed: int = 123546, trials: int = 100, verbose: int = 0):
```
to speed this up, can we please lower the trials to 10, and not 100

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+from textwiser import TextWiser, Embedding, PoolOptions, Transformation, WordOptions
+
+
+
```
Per PEP-8 standard, we use 2 blank lines between Class definitions and 1 blank line between methods. If you are using PyCharm it should be underlining these python standard errors

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+        labels = pd.DataFrame({"item1": [0, 1, 0, 0, 0], "item2": [1, 0, 0, 1, 0], "item3": [0, 0, 1, 0, 1],
+                               "item4": [0, 1, 1, 0, 1], "item5": [0, 1, 0, 0, 1]})
+
+        print(data)
```
I assume these are for debugging for now.. in general, when runnign the code we should not get any prints

skadio replied:
same comment in other tests.

zohrehraziei replied:
Yes, I added them for debugging purposes. I removed the print(data) and print(labels). Line selector.transform(data), the transform used only for printing selected features since decision makes in fit. Any tips on what the best functionality is for the transform function?

skadio replied:
That is a GREAT question! @takojunior  @bkleyn  for visibility. 

We talked about different Text_Based methods, which are essentially captured in the rows of your Slide, but did not talk about what fit() and transform() means for each method. 

That's a bit of an open ended design decision. 

For example, 
-  In one extreme, we can say transform() does nothing and fit() implements the methods in its entirety. 
- In the other extreme, transform() does all the job and fit just returns the selected columns. 

A hybrid might be possible; that is; the work for the method is split across Transform() and Fit(). The guiding principle being, transform does the necessary job/transformation for the method to work, and fit() does the main bulk of the work. 

For example, for random() if a K is given, transform() does not need to anything and fit() return K random columns. IF a K is not given, then transform() can figure out what the K is usign unicost or diverse as desired. Essentially, it "prepares" what's all needed to make the final selection. Then fit() again returns k random columns. 

Whatever "design decision" we make here, we NEED to be consistent across all rows in the Slide that the idea behind the Transform() and Fit() logic is the same. So we might ne able to decide it solely based on looking at Random() 

Thoughts from others?

takojunior replied:
I feel the hybrid approach is reasonable. But I try to have one step back. For random(), and other methods, we are supposed to run fit() first and then transform() right? Or are we allow user to call transform() directly? I feel that if fit() already solves the selection by returning k columns (when k is not set, it can be figured out by solving the set_cover problem), then the transform() might just use the k returned indices to extract the data?

skadio replied:
You'r right.. in my comment I made a mistake in the order of FIT + TRANSFORM. Let me edit my comment to avoid confusion

skadio replied:
Comment updated, it should follow FIT + TRANSFORM sequence now.

zohrehraziei replied:
In the current implementation, the fit() function selects the desired number of text-based features using a content selection method and stores the selected feature names in the selected_features_ attribute. Then, the transform() function extracts these selected features from the input data and prints them to the console before returning them as a pandas DataFrame. This approach ensures that the selected features are consistently used for both training and inference.
Shall we continue with this approach?

skadio replied:
Sounds reasonable to me.. Let's proceed with as-is now.. and have a mental note for it ONCE we are done with all the methods and then we can see if FIT() + TRANSFORM() operations are somewhat meaningful/consistent across methods.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+
+        selector = Selective(method)
+        selector.fit(data, labels)
+        selector.transform(data)
```
can we add some self.asserTrue() to check these outputs. Currerntly it just runs but does not verfiy anything in the output. See other tests in other classes and things they verify as the output

skadio replied:
same comment for other tests

zohrehraziei replied:
Sure, I added some assertations to verify that the output of the tests are correct and consistent.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+        selector.fit(data, labels)
+        selector.transform(data)
+
+
```
extra lien

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+        selector.fit(data, labels)
+        selector.transform(data)
+
+
```
extra lien

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
```
Can you confirm that all other tests are passinng? To make sure that our initial update for Text_Based is not breaking anything else. If you right-click on Tests folder, you can run unittest, and it runs everything

skadio replied:
Yay! :D

bkleyn replied:
I'm not able to run the tests. Some are passing, but mostly failing. Guessing this is due to recent changes.

zohrehraziei replied:
You should be able to run the tests in the new commits. Please let me know if you have any issue.

skadio reviewed (approved):
Done with a quick pass, looks good to me. I tried to keep my comments for "random" related parts, but a few other comments here and there. 

For random, 
- see the opportunity for creating a test for _calculate_size(cost_method) that we can reuse. 
- verify test behavior, not just run them 
- fix the seed issue
- expose trials to the main method 
- other minor comments inlined

skadio commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+            * random: This method finds a random solution by selecting a random set of items.
+                           The number of selected times (num_features = t):
+                                - t = maximum number of features defines by user.
+                                      The cost_metric input argument needs to ignore in this case.
```
may be rephrasing here; "The cost_metric input argument is ignored in this case"

zohrehraziei commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+                - Greedy optimization with unicost as cost_metric
+
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
```
I have changed the code for random selection. Yes, I need to change the pydoc based on the new code. The code is selecting a random sample of columns from the dataset self.matrix of size selected_size without repetition for a fixed number of trials. 
Yes, it is a great idea! I passed  "num_trials" (named it "trials") as an argument.

In the pydoc, I explained that the "trials" parameter controls how many trials are performed during the random selection process and that the function returns the best random selection in terms of label coverage out of the specified number of trials.

skadio replied:
Now I am confused actually.. Selecting the best out of num_trial random sounds good to me. I just meant the user is not aware of it, which we can mention in the pydoc (in selector.py) 

You said, the code has changed now. What's the change compared to previous random() code/algorithm?

zohrehraziei replied:
The random selection original code was _select_random(self, selection_size: int)->List. It uses the random.sample() method to return a list of `selection_size' number of random non-repeating items selected from a range of indices.

The current random selection function, for trials number of iterations, randomly selects a sample of selected_size items from a range of indices, making sure that there is no repetition in the selection. It calculates the number of covered categories for each random selection, and keeps track of the one with the highest coverage. Finally, it returns the list of indices with the highest coverage.

My intuition from your comment was that you asked to include the number of trials as an additional optional parameter when calling the TextBased method in test_text.py. So, we can set the default value for trials to 10 and for random selection to any value the user wants.

Below is an example:

method = SelectionMethod.TextBased(num_features=3, optimization_method="random", cost_metric=None, trials=10)

Please correct me if I thought wrongly.

skadio replied:
I am on the same --your example makes perfect sense! 

Wasn't the original _select_random() code also picking out the "Best" out of num_trials too?

zohrehraziei replied:
No, the original code does not involve the best selection out of multiple trials. It simply selects selection_size number of columns randomly from the available columns.

zohrehraziei replied:
@takojunior 
 Based on this comment, please read the changes made to the number of trials. Please let me know if it answers your question about the number of trials.

skadio replied:
I see now. thanks for the clarification!

takojunior replied:
> 

Thanks for the clarification!

takojunior reviewed (approved):
I did my first pass. I left a few comments to focus on the random() selection for now as I think we are seeing it as the first stage. Those comments are mainly for parameterizing `trials` in _TextBased class, and understanding better how random() behaves in different cases when `num_features` is set or not set, combined with `unicost` or `diversity_cost`.

Before merging, please clarify the questions and make changes, for example the places pointed out for coding styles, e.g. drop some blank lines.

zohrehraziei commented on feature/text_based.py:
```diff
@@ -0,0 +1,724 @@
+        self.featurization_method = featurization_method
+        self.optimization_method = optimization_method
+        self.cost_metric = cost_metric
+        self.content_selector = ContentSelector(selection_size=num_features, seed=123456, trials=100, verbose=True)
```
Thank you for mentioning it! Done!

zohrehraziei commented on tests/test_text.py:
```diff
@@ -0,0 +1,223 @@
+        print(labels)
+
+        method = SelectionMethod.TextBased(num_features=3,
+                                           optimization_method="random")
```
Please refer to the comments I tagged you

zohrehraziei commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+        featurization_method :
+            TextWiser object to featurize items in X_textual (it is named data in test and this method does not
+            have impact on the following options:
+                - Random optimization with num_features = max_num_features
```
The Pydoc documentation is currently not up-to-date for all optimization methods. However, I will update it for each method as soon as I create a pull request for it.

Regarding the "max_num_features" parameter, I have made the decision to change it to "num_features". This is because this parameter can now be set to any number less than or equal to the maximum number of features that the user is able to select.

zohrehraziei commented on feature/selector.py:
```diff
@@ -270,6 +273,102 @@ def _validate(self):
+                        num_feature must be an integer for third optimization level (solving max cover).
+
+        cost_metric : str, optional;
+            * unicost: Each item/feature incurs a cost of one when included in the selection.
```
If the "num_features" parameter is set to "None", we will need to solve a set cover problem to determine the selection size. If the cost metric is not defined, a set cover problem will be solved by default unicost. However, if the cost metric is set as "diverse", then a set cover problem with a diverse cost in the objective function will be solved to determine the selection size.

skadio replied:
exactly! 
@takojunior  take a look at the Slide Zohreh shared. Each configuration of params is explained in the Table about what they do (num_features X optimization_method X cost_metric configurations)

zohrehraziei replied:
I need to just make a correction. The default is diverse not unicost. So, it cost metric does not define, it will solve a diverse set cover problem

zohrehraziei commented:
Hi,

This pull request also includes the Greedy optimization method with the options mentioned in the second row of Tables 1 and 2, namely greedy_unicost(t) and greedy_diverse(t), as well as their variations greedy_unicost(t=unicost()) and greedy_diverse(t=diverse()).

It's worth noting that no new functions have been added to text_based.py aside from the function for greedy selection, which is _select_greedy(). Please review this function from the ContentSelector class and the required modifications based on the first pull request.

Please also, review the README.md file which includes a test example with data/labels as a section after benchmarking section.

Thank you for your review.
Zohreh

zohrehraziei commented:
Hi,

This pull request also includes the KMeans optimization method with the options mentioned in the third row of Tables 1 and 2.
The new functions have been added to text_based.py include  _select_kmeans() and _get_closest_to_centroids().

Thank you for your review.
Zohreh

zohrehraziei commented:
Hi,

This pull request also includes the Exact optimization method with the options mentioned in the fourth row of Tables 1 and 2.
The new functions have been added to text_based.py include _select_multi_level_optimization() and _solve_max_cover().

Thank you for your review.
Zohreh

skadio commented on README.md:
````diff
@@ -102,6 +104,67 @@ stats_df = calculate_statistics(score_df, selected_df)
 ```
 
+## Example of Text-based Methods
+The **TestBased** selection method can be used to select the reviews that most representative of the overall sentiment
````
typo; textbased not testbased :D

skadio commented on README.md:
````diff
@@ -102,6 +104,67 @@ stats_df = calculate_statistics(score_df, selected_df)
+```python
+import pandas as pd
+from feature.selector import Selective, SelectionMethod
+from tests.test_base import BaseTest
````
why import Test?

zohrehraziei replied:
Removed, it was required in class TestText

skadio commented on feature/selector.py:
```diff
@@ -270,6 +269,129 @@ def _validate(self):
 
+    class TextBased(NamedTuple):
+        """
+        # TODO  review the content to double check everything is %100 correct once ALL the implementation is done
```
please revise

skadio replied:
let's get this into its final shape now..

zohrehraziei replied:
I re-check everything in PyDoc, just not sure about the following TODOs:
1) If no Y is given, it selects #TODO..
2) Randomness: Behavior for #TODO methods is non-deterministic, depends on seed.
for the second one, I explained what is the source of randomness in each optimization methods. I am not sure about the first one.

skadio commented on feature/selector.py:
```diff
@@ -381,14 +504,19 @@ def __init__(self, selection_method: Union[SelectionMethod.Correlation,
             raise ValueError("Unknown Selection Method " + str(selection_method))
 
-    def fit(self, data: pd.DataFrame, labels: Optional[pd.Series] = None) -> NoReturn:
+    def fit(self, data: pd.DataFrame, labels: Optional[Union[pd.Series, pd.DataFrame]] = None) -> NoReturn:
```
I don't think we need Union inside Optional. Union creates a new type, no? This can be Optional[Series, DF] 

Please also make the change in the fit_transform() functions.

zohrehraziei replied:
This change causes to get error message: "typing.Optional requires a single type. Got (<class 'pandas.core.series.Series'>, <class 'pandas.core.frame.DataFrame'>)."

This is due to the fact that 'typing.Optional' only accepts a single type as argument. 'Optional' required single type, but we passed two types (pd.Series, pd.DataFrame) to fit and fit_transform. So, it is better to use Union to specify that both types are allowed.
I think it is also the reason that @bkleyn gets error for passing the tests. 
Shall we continue with having  `Union'?

skadio replied:
I see.. Possible that I forgot how Typing works and it seems we "have to" put a Union[x, y] inside the Optional. @bkleyn 

IF this is the case, can you make sure this is consistent across all FIT(), TRANSFORM() calls all the way from public api in selector.py to under the hood in TextBased

bkleyn replied:
I think we have to use Union. 

@skadio you are right that it effectively creates a new type. This is what we have in other libs as far as I can see.

skadio replied:
cool! thanks for raising @zohrehraziei  and thanks for confirming @bkleyn

skadio commented on feature/selector.py:
```diff
@@ -270,6 +269,129 @@ def _validate(self):
+       trials: int;
+            The number of random trials to perform. Defaults to 1.
+        """
+        # Defualt values
```
typo defAult

skadio commented on feature/selector.py:
```diff
@@ -448,14 +577,21 @@ def _validate_args(seed, selection_method) -> NoReturn:
 
     def _validate_fit(self, data, labels):
 
+        check_true(data is not None, ValueError("Data cannot be none"))
```
great catch!

skadio commented on feature/selector.py:
```diff
@@ -448,14 +577,21 @@ def _validate_args(seed, selection_method) -> NoReturn:
+
+                # For text_based labels is a dataframe
+                if isinstance(self._imp, _TextBased):
+                    check_true(isinstance(labels, pd.DataFrame), ValueError("Labels should be a pandas dataframe."))
```
you can also assert that the number of columns in "data" and "labels" should match each other.

zohrehraziei replied:
Added!

skadio replied:
I think I dropped a comment related in the Tests, that, we can check that invalid sizes between data and label is handled nicely

zohrehraziei replied:
Added!

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+
+    @staticmethod
+    def _verify_input(cost, matrix):
+        check_true(cost is not None, "Data Error: cost cannot be none")
```
the second arguments is not text.. it needs to be an "exception". 

curious about your PyCharm setup; don't you see large marks of yellow errors coloring text here for these issues?

zohrehraziei replied:
Changed it with 'ValueError' to indicate that there is an issue with the values being passed to the function.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+# TODO not sure if this should use _BaseDispatcher or not. We can decide later
+class _TextBased(_BaseSupervisedSelector):
+    """
+    featrization and optimization are done in 'fit' function.
```
typo featUrization

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        self.cost_metric = cost_metric
+        self.trials = trials
+        self.content_selector = ContentSelector(selection_size=num_features,
+                                                seed=self.seed, trials=trials, verbose=True)
```
why is verbosity True by default? Shouldn't this be False?

zohrehraziei replied:
Yes, set it to False to reduce the output printing

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        super().fit(data, labels)
+
+        # Get the text columns dynamically
+        text_columns = [col for col in data.columns if col.startswith("item")]
```
what does this line do? And how can we select columns with a hardcoded column name? How can we guarantee columns will start with "item" string?

skadio replied:
Then on 615 we are turning columns again to subset with "selected" only, so cannot we loop df.columns on line 615 and remove this line?

zohrehraziei replied:
Removed line 608, and directly looped it through the data.columns

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        # Get the text columns dynamically
+        text_columns = [col for col in data.columns if col.startswith("item")]
+
+        selected_indicies = self.content_selector.run_content_selection(data, labels, self.featurization_method,
```
Help me follow something here.. the "labels" that came to this method is a DF, yes? 

We are sending this DF to the run_cs() method here. But the second argument of run_cs() is expecting a List[int]. so how does this work?

zohrehraziei replied:
Modified labels as DF. But as long as the labels has correct shape and dtype (means DF) for the selection process the previous code works.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+    def transform(self, data: pd.DataFrame) -> pd.DataFrame:
+
+        feature_selected = data[self.selected_features_]
+        print("Selected items:")
```
we don't want prints in the library

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+
+        # Only select the features that were selected during fit
+        selected_features = [col for i, col in enumerate(text_columns) if i in selected_indicies]
+        self.selected_features_ = selected_features
```
So we are adding a new field called selected features in the init() and then use it here.. which makes total sense

This code is fine as-is but within this library.. but to be consistent with every other method.. maybe we can drop this variable and instead use the abs_score field from the base..

In the public api, we have a get_abs_score() method. We need to come up with some "score" to return for each feature.. For instnance, a linear selection can return regression coefficients as the "abs_score". 

I don't see any clear definition of a "score" for TextBased. We can treat score as 0/1 binary, set it to score=1 if selected and 0 otherwise. Setting abs_score to 1 using selected_indices this way replaces the selected_features variable here

We can also set self.num_features = len(selected_indices) (IN CASE, we end up selecting less features/items than asked)

Then in transform() we can make a call to get_top_k() like in other methods, and it returns top_num_features from abs_scores --which should return the selected indices without tracking a separate "selected_features" 

As a bonus, the get_abs_score() method in the public API also works for TextBased.

zohrehraziei replied:
Great Idea! 
Made following changes:
Instead of using the selected_features variables to store the selected columns, I set the absolute score for each feature in the fit() method and storr the abs_scores field. In transform() method, the top_k selected features will return when the num_features = len(selected_indicies)

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        """
+
+        # Process data
+        self.cost_metric = cost_metric
```
when setting these variables, it is better to keep their ordering, in the same order as the function signature to avoid confusion

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        List of indices in data that are selected.
+        """
+
+        # Process data
```
can we add a word here what does process data does

zohrehraziei replied:
Please see to the next comment. I also added more comments in the code.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+    def _process_df(self, input_df: pd.DataFrame, categories: List[int], selection_size: int,
+                    optimization_method: str, cost_metric: str) -> NoReturn:
+        """_process_df
+        We have implemented two versions of this function. In the version here, the categories parameter
```
aren't we removing the other one? we don't need both correct?

skadio replied:
Can we add some explanation here what this process is doing. I am not following it

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        """
+        self.matrix = categories.to_numpy()
+        self._num_rows, self._num_cols = self.matrix.shape
+        option_map = {
```
what is this options map? Can we add some comments to this blocks of code plz

zohrehraziei replied:
Added some comments for options map

skadio replied:
OK. I spent quite a bit of time to understand this.. becase there is a lot happening.. there maps, lambdas, factory methods etc. 

IF I am correct, all we ware trying to do here is to decide whether we need text featurization or not. YES? 

IF so, you can very much simplift this (e.g., you don't need lambdas that are none jsut to pass and a bit of more explicit naming) 

```
from collections import defaultdict

is_featurization_needed = defaultdict(True) 
is_featurization_needed[("random", True, "diverse")] = False
is_featurization_needed[("random", False "unicost")] = False
is_featurization_needed[("random", True, "unicost")] = False
is_featurization_needed[("greedy", False "unicost")] = False
is_featurization_needed[("greedy", True, "unicost")] = False
is_featurization_needed[("exact", True, "unicost")] = False
is_featurization_needed[("exact", False "unicost")] = False

if is_featurization_needed: 
   feature_column = self.featurization_method.fit_transform(input_df)
````

skadio replied:
why is not (random, False, diverse) is not added as False?

skadio replied:
I am also thinking about converting all True -> False and False -> True in the is_featurization_needed dictionary 
so that the configuration can be as (opt_method, selection is None, cost_metric) to avoid the negation in the middle. 

Sth to consider, if you believe it makes it easier to parse/reason about?

skadio replied:
Higher level comment about this process_data function. I don't think it is needed. Back then, when dealing with categories, some processing was involved. Here the input "labels" are already in the format we want. 

So this part can totally be moved into the constuctor and removed from this function:
```
        self.matrix = labels.to_numpy()
        self._num_rows, self._num_cols = self.matrix.shape
```

THEN.. what remains here is text featurization. So I would turn this function around in the constructor and make it 

```
self.text_embeddings = _get_text_embeddings(selection_size, optimization_method, cost_metric) 
```

(notice I renamed features to text_embeddings.

zohrehraziei replied:
I agree with overall idea, expect these two points:

1- Move below part to constructor:
self.matrix = labels.to_numpy()
self._num_rows, self._num_cols = self.matrix.shape

If we move the code for setting self.matrix and self._num_rows and self._num_cols into the constructor, then we need to make sure that these values are set before calling _get_text_embeddings in the constructor. As a result, the inputs to the _TextBased class constructor will need to change to include labels, and any code that creates an instance of the _TextBased class will also need to include labels.
 Also, if we move the process of generating text embeddings to the constructor, we would need to add input_df as an input argument to the SelectionMethod.TextBased() constructor. This would also require adding input_df and labels as inputs to the ContentSelector class. This change would break the consistency of input arguments across different selective methods.

Therefore, moving those lines to constructor require changing the input arguments of the SelectionMethod.TextBased() constructor and adding input_df and labels as inputs to the ContentSelector class. It would also make storing labels and the number of rows and columns in the constructor unnecessary.

2- (random, False, diverse) -> rand(t=diverse()) when t=None. It mean solving a set cover with diverse cost which it requires featurization.

skadio replied:
#2 good there, thanks for clarifying. 

#1, I made a mistake. I didn't mean the constructor but lifting those lines from process() to where process() is called, which is run() (i thought we were in the init().. but I meant run()

So like this;
```
def run()
   ... 
    self.matrix = labels.to_numpy()
    self._num_rows, self._num_cols = self.matrix.shape
    self.text_embeddings = _get_text_embeddings(selection_size, optimization_method, cost_metric)
    ... 
````
This is the "same" as before except that process() can be renamed to get_text_embeddings() now and only does that. In fact there is no real processing anymore (unlike the category stuff before) we are just reading the matrix from labels

This simplifies/clarifies

zohrehraziei replied:
Make sense. Thank you for clarification! Done!

skadio replied:
Cool. I assume you will push an updated version soon

zohrehraziei replied:
Yes, I will do soon

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+            ("exact", False, "unicost"): lambda: None
+        }
+
+        options = option_map.get((optimization_method, selection_size is not None, cost_metric))
```
curios what's happening here. also, the naming makes this a bit obscure. both the IF-condition is options, and then IF options then it calls options ..

zohrehraziei replied:
We are using a dictionary called option_map to map a specific set of arguments (optimization_method, selection_size is not None, and cost_metric) to a corresponding lambda function. If an entry in the dictionary matches the given argument set, then the corresponding lambda function is retrieved and called using options(). If not, the code proceeds with feature generation and matrix creation as normal. This part of code provides a easy and efficient way to handle different cases without having to write a lot of if/else statements.

zohrehraziei replied:
Some comments also added to this part of code for clarification.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+            feature_column = self.featurization_method.fit_transform(input_df)
+            self.features = np.array([eval(l) if isinstance(l, str) else l for l in feature_column.tolist()])
+            check_true(len(self.features) == self._num_cols,
+                       f"Process Data Error: features size ({len(self.features)}) "
```
these should not be strings.. the second argument is an Exception.. 

That said,  these conditions should NOT occure, correct? The code should never come here. 

We are done with all our validate_args(), validate_fit() functions etc. so by the time the code is deep in the library all should work.. Anytime it does not, it is a because of a bug or sth  --not because expectations are not met. 

Agree that code should not hit these invalid conditions?

zohrehraziei replied:
Yes, I agree that the code should ideally never hit these invalid conditions. However, it might be good to include such checks using raise ValueError as it is clearly indicates that an error has occurred and provides a meaningful error message. 
Please let me know if I need to remove it.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        # Return
+        return selected
+
+    def _solve_set_cover_cost_option(self, cost: np.ndarray) -> List:
```
I am not a big fan of this function.. The reason we have the DATA class is not pass back and forth individual elements. I would rather replace Data.cost outside and just call solve_set_cover(data) and not have this function 

(Notice the "implicit" use of "matrix" inside which is very obscure from outside to understand what is this solving)

zohrehraziei replied:
The function has been removed. The reason of adding this function was in _select_multi_level_optimization(), we need to do Data.cost and call solve_set_cover(data) multiple times for if/else statements.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+
+    def _select_multi_level_optimization(self) -> List:
+        unicost = np.ones(self._num_cols)
+        if self.selection_size is None:
```
cannot use get_selection_size()

zohrehraziei replied:
This comment is not clear to me. Would you please explain more?

skadio replied:
I meant cannot we re-use get_selection_size() here

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+    def _get_diversity_cost(self, selected_size: int) -> List[float]:
+        """
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
```
@bkleyn interesting that we didn't run into this before.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        """
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
+             costs become zero. There are two ways to handle this situation:
```
can you add a comment on why the code picks or the other

zohrehraziei replied:
The reason why we need to add NEXT_S_DIS or DUMMY is that there are two alternative options to handle situations where the minimum distance between each two features is zero. This situation may occur if the features are very similar or if there are a limited number of features. When this happens, the diversity cost becomes zero. Therefore, the code adds NEXT_S_DIS to replace the zero value in the distances matrix with the next smallest non-zero distance. This ensures that each feature has a non-zero diversity cost, which helps to select a more diverse set of features even when the minimum distance between each two features is zero. Also, DUMMY may not be a good option (it is @takojunior suggestion) for Greedy since the adjusted cost brings all values between (0, 1). In this case, there is not much difference between uni-cost and diverse cost, since the adjusted cost elements are so similar. We need to decide with which option we want to continue.
I also copied this comment for the same question @bkleyn asked.

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        return diversity_cost
+
+    def _get_num_row_covered(self, selected) -> int:
+        check_true(self.matrix is not None, "matrix is not initialized")
```
second param is an Exception

zohrehraziei replied:
Edited

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+        # Solve (optimize using Gurobi)
+        model.verbose = 0
+        model.optimize()
+        check_true(model.status == OptimizationStatus.OPTIMAL, "Max Cover Error: optimal solution not found.")
```
second parameter is Exception

zohrehraziei replied:
Edited

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,694 @@
+    def _solve_max_cover(self, data: Data, selected: List) -> List:
+        # If selected is given, limit the max_cover_size
+        if selected is not None and self.selection_size is not None and len(selected) > 0:
+            assert (self.selection_size <= len(selected)), "Max Cover Error: max_cover_size cannot exceed num selected"
```
remove assert, and make sure to use Exception

skadio replied:
actually I am not sure about this assert. say more why is this needed.

zohrehraziei replied:
The assert statement in the _solve_max_cover function was appropriate for the old structure of the ContentSelector class. However, in the new MaxCoverSelector class, this condition is already checked before calling the function, so the assert statement is redundant and has been removed in the new commit.

skadio reviewed (commented):
@zohrehraziei  I am done with another pass. Happy to look at it more. I left very detailed comments. There are some gray areas (like process_data()) in my mind, and overall let's clean-up all these to bring this to the final stage.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        selector_diverse.fit(data, labels)
+        selected_features_diverse = set(selector_diverse.transform(data).columns)
+
+        assert selected_features_unicost == selected_features_diverse
```
asserts are not used in tests. replace them with self.assertTrue() from the unittest framework. Same comment for all tests

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        selector_diverse = Selective(method_diverse)
+        selector_diverse.fit(data, labels)
+        selected_features_diverse = set(selector_diverse.transform(data).columns)
+
```
If you could add a quick comment above the assertion about the exception, it makes it a lot easier to understand what the test is doing. For example, here it would be sth like "random should result the same selection with uni/diverse cost"

zohrehraziei replied:
Added.

zohrehraziei replied:
A comment added for each test

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+
+        # Check whether the number of columns in `selected_features' is equal to num_features (user input) or not
+        if method.num_features is None:
+            pass
```
so this test is just passing? (num_features is given as None above)

zohrehraziei replied:
Modified all tests for this assert.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        labels = pd.DataFrame({"item1": [0, 1, 0, 0, 0], "item2": [1, 0, 0, 1, 0], "item3": [0, 0, 1, 0, 1],
+                               "item4": [0, 1, 1, 0, 1], "item5": [0, 1, 0, 0, 1]})
+
+        # "diverse" is Default cost metric
```
better to set diverse explicity in "tests". IF default changes tomorrow, this test will fail

zohrehraziei replied:
Added.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+                               "item4": [0, 1, 1, 0, 1], "item5": [0, 1, 0, 0, 1]})
+
+        method = SelectionMethod.TextBased(num_features=None,
+                                           optimization_method="random",
```
don't you wanna set seeds when random is used? Same comment in all tests with random

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+
+        self.assertTrue(isinstance(selected_features, pd.DataFrame))
+        self.assertEqual(selected_features.shape[0], data.shape[0])
+        if method.num_features is None:
```
I think this test also just pass-es and not verifies

zohrehraziei replied:
New asserts added to verify the selection

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        assert selector.selection_method.trials == 10
+        self.assertTrue(isinstance(selected_features, pd.DataFrame))
+        self.assertEqual(selected_features.shape[0], data.shape[0])
+        if method.num_features is None:
```
why do we need if-else here? can this ever be none, set as 2 above, no?

zohrehraziei replied:
Modified (it is not if-else anymore)

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        if method.num_features is None:
+            pass
+        else:
+            self.assertEqual(selected_features.shape[1], 2)
```
We need to verify the selected items in these, no? Not just the size? Cannot you setup the labels columns such that we know the greedy solution clearly, and then assert the selections here. 

Immediate exampels would be, a greedy method with Idenitiy labels -- it should select all. 
Another is greedy with a single coluimn covering everying --it should definitely select that index 

And what happens with greedy when 1) there is no label coverage at all 2) some labels cannot just be covered (no item has it, the instance is infeasible)

zohrehraziei replied:
Please check the tests. The mentioned tests are added for the tests based on their relevance.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+                                           featurization_method=TextWiser(Embedding.TfIdf(min_df=0),
+                                                                          [Transformation.NMF(n_components=20),
+                                                                           Transformation.SVD(n_components=10)]),
+                                           optimization_method="kmeans",
```
wjhen it is kmeans we probably wanna set seed?

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+        self.assertEqual(selected_features.shape[0], data.shape[0])
+
+    ################################################
+    ########## Verify invalid tests  ###############
```
how about size mismatch between data and labels?

zohrehraziei replied:
Added

zohrehraziei replied:
The related assert added

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,464 @@
+from textwiser import TextWiser, Embedding, Transformation
+
+
+class TestText(BaseTest):
```
Lots of good tests here! My only high-level comment is
1) not following some of the if-else none pass code
2) we are not verfiying the selection. It would be good to capture the exact selected items so that we can reproduce the tests identically. Notice the size of the data is small 4 columns or so, hence returning a data with 2 subsets is very likely evevn if the 2 selected is "incorrect" 

For Random, we can assert to select the same indices with the same seed. You can actually push a "best" solution, and even its random, it should end up selecting  the best index 

For Greedy, you can constrcut labels such that you know the greedy columsn. Verfiy that they are selected at the end. 

For Kmeans, it is not obvious to me how to do that index verifications. I would just run it once, and then assert the same selection from this moment onward (with the same seed). 

For Exact, same idea. run once and then assert the same selection of indices going forward. 

BTW, are we sure to test most (if not all?) combinations? I know there are too many parameterization with/without features (2) X num_opt (4) X num_cost 2 = 16. EVEN IF we are not asserting anything, just running each combination at least ONCE is a test in and of itself.

zohrehraziei replied:
Regarding the if-else none pass code, I made changes to simplify the code to make it more readable and easier to understand. 

I agree that it would be good to capture the exact selected items so that we can reproduce the tests identically.
For Random, I have added an assert statement to verify that the same indices are selected with the same seed. I also added a comment suggesting that for the "random" optimization method, we can push for a "best" solution, and even though it is random, it should end up selecting the best indices. 

For Greedy, I have constructed labels to know the greedy columns and verified that they are selected at the end. We should set the seed for Greedy as there is a source of randomness in the initialization of the Lagrangian multiplier.

For KMeans and Exact, we have run the tests once and verified that the same selection of indices is chosen going forward with the same seed. For Kmeans, we need to set the seed as follows: 
seed = seed value
np.random.seed(seed)
random.seed(seed) 
sklearn.set_config(working_memory=0) sklearn.utils.check_random_state(seed)
The reason is that Kmeans uses random initialization of cluster centroids. Even with the same random seed, the starting position of centroids can differ due to various factors such as the optimization algorithm used, data normalization, etc. I set the random seed of NumPy and scikit-learn packages to ensure reproducibility. This will ensure that the results are reproducible across multiple runs.

I have tested all possible combinations of features, optimization methods, and cost metrics to ensure that all combinations are tested at least once.

skadio replied:
"We should set the seed for Greedy as there is a source of randomness in the initialization of the Lagrangian multiplier." 

Is this captured in the PyDoc of TextBased in selector.py

zohrehraziei replied:
Added to the PyDoc for greedy

bkleyn commented on feature/text_based.py:
```diff
@@ -0,0 +1,702 @@
+        # Get the text columns dynamically
+        text_columns = [col for col in data.columns if col.startswith("item")]
+
+        selected_indicies = self.content_selector.run_content_selection(data, labels, self.selection_size,
```
The type hinting says labels should be a list, but here we are passing a data frame?

zohrehraziei replied:
Solved this issue. Please review the code and let me know if you have comment.

bkleyn commented on feature/text_based.py:
```diff
@@ -0,0 +1,702 @@
+        best_covered = 0
+
+        if self.selection_size is None:
+            self.selection_size = self._get_selection_size()
```
_get_selection_size() returns a tuple so I'm not following logic in line 449. Did you intend to number of items selected (i.e., second value in tuple)?

skadio replied:
yeah this is an error (I made it)

zohrehraziei replied:
I fixed and you should be able to run all the tests.

bkleyn commented on feature/text_based.py:
```diff
@@ -0,0 +1,702 @@
+import numpy as np
+from feature.base import _BaseSupervisedSelector
+from feature.utils import Num, check_true
+import matplotlib.pyplot as plt
```
Unused import

bkleyn commented on feature/text_based.py:
```diff
@@ -0,0 +1,702 @@
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
+             costs become zero. There are two ways to handle this situation:
+                * add dummy cost to diversity cost (DUMMY)
```
Looks like the dummy cost block has been commented out / removed?

zohrehraziei replied:
The reason why we need to add NEXT_S_DIS or DUMMY is that there are two alternative options to handle situations where the minimum distance between each two features is zero. This situation may occur if the features are very similar or if there are a limited number of features. When this happens, the diversity cost becomes zero. Therefore, the code adds NEXT_S_DIS to replace the zero value in the distances matrix with the next smallest non-zero distance. This ensures that each feature has a non-zero diversity cost, which helps to select a more diverse set of features even when the minimum distance between each two features is zero. Also, DUMMY may not be a good option (it is @takojunior suggestion) for Greedy since the adjusted cost brings all values between (0, 1). In this case, there is not much difference between uni-cost and diverse cost, since the adjusted cost elements are so similar. We need to decide with which option we want to continue.

bkleyn commented on tests/test_tree.py:
```diff
@@ -14,7 +14,7 @@
 from feature.selector import Selective, SelectionMethod
 from feature.utils import get_data_label, Constants
-from tests.test_base import BaseTest
+from .test_base import BaseTest
```
Why this change? All the other test modules use `from tests.test_base ...`

bkleyn commented on tests/run_all.py:
```diff
@@ -19,4 +19,4 @@
 
 # Run the Suite
-runner.run(suite)
+runner.run(suite)
```
Add back whitespace

bkleyn commented on feature/utils.py:
```diff
@@ -78,6 +78,36 @@ def normalize_columns(df: pd.DataFrame):
     return df_normalized.fillna(0)
 
 
+def _process_data(input_df: pd.DataFrame, categories: List[str], feature_column: str, selection_size: int) \
```
This isn't being used anymore? I'm not following the relationship between this function and the _process_data() in text_base.py. It looks like they are doing different things though.

skadio replied:
commenting this out for now to avoid confusion (and taking back into text_based.py file)

bkleyn commented on feature/selector.py:
```diff
@@ -12,8 +12,7 @@
 
```
Can you please delete lines 7-8? (author and version captured elsewhere)

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+
+        if self.selection_size is None:
+            if self.cost_metric == "unicost":
+                selected = self._solve_set_cover(data)
```
don't you need to set k here ?

zohrehraziei replied:
No, we do not need it. 'k' is only required when num_features is not None. For these two conditions

Punicost=unicost().​
If t <unicost_selection,
max_cover(Punicost, t)

AND

k=unicost ().​
diverse_costs=kmeans(k). ​
P_diverse=diverse().  ​
If t <diverse_selection, 
max_cover(P_diverse, t).

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+        self.selection_size = selection_size
+        self.seed = seed
+        self.trials = trials
+        self.verbose = False
```
shoudl not this be set to verbose? from the constructor

zohrehraziei replied:
solved!

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+        super().__init__(seed)
+
+        # Set input arguments
+        self.num_features = num_features
```
why do we have both self.num_features and self.selection_size? Cannot remove one (say selection_size)

zohrehraziei replied:
Yes, we can. Removed selection_size and replaced with num_features

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+        # Call fit method from parent class _BaseSupervisedSelector
+        super().fit(data, labels)
+
+        selected_indicies = self.content_selector.run_content_selection(data, labels, self.selection_size,
```
self.content_selector.run_content_selection(data, labels, self.selection_size,) 

why are sending in the self.selection_size and not the self.num_features?

zohrehraziei replied:
They are the same. We kept num_features

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+            feature_column = self.featurization_method.fit_transform(input_df)
+            self.features = np.array([eval(txt_feature) if isinstance(txt_feature, str)
+                                      else txt_feature for txt_feature in feature_column.tolist()])
+            if len(self.features) != self._num_cols:
```
can we please turn these a if-raiseError statements into checkTrue() 

We should always use checkTrue() to raise these asserts.

skadio replied:
(don't forgot to negate the conditions as required when replacing with checkTrue)

zohrehraziei replied:
Done! I used check_true(true condition, ValueError())

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+    def fit(self, data: pd.DataFrame, labels: pd.DataFrame) -> NoReturn:
+
+        # Call fit method from parent class _BaseSupervisedSelector
+        super().fit(data, labels)
```
I don't we need to run this..

zohrehraziei replied:
You are right. We already called _BaseSupervisedSelector's fit() method by calling super().__init__(seed) in _TextBased's constructor

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+        self._num_rows = None
+        self._num_cols = None
+
+    def run_content_selection(self, input_df: pd.DataFrame, categories: pd.DataFrame, selection_size: int,
```
As far as I understand, we are not working with "categories" anymore (hence, that's why we changed the process_data method here). Therefore, this parameter should not be called "categories" here. Please also fix the pydocs and drop categories terminology from everywhere

zohrehraziei replied:
I replaced all of them with labels and fixed pydocs

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+    matrix: np.ndarray
+        Coverage matrix where each row corresponds to all predefined labels for each category and the columns
+        correspond to the content.
+    features: np.ndarray
```
How about we replace "features" -> "text_embeddings".. Selective is a feature selection library so calling the text embeddings here as "features" is going to be confusing.

zohrehraziei replied:
Replaced "features" with "text_embeddings".

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+
+        return selected
+
+    def _process_data(self, input_df: pd.DataFrame, categories: pd.DataFrame, selection_size: int,
```
I guess we should not call this input parameter as "categories" anymore?

zohrehraziei replied:
Fixed

skadio commented on feature/text_based.py:
```diff
@@ -0,0 +1,703 @@
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
```
I don't think there is any categories involved here?

zohrehraziei replied:
Fixed

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,775 @@
+        selected_features_diverse = selector_diverse.transform(data)
+
+        # Check whether the seed value used in tests is the same (not the default = 123456)
+        self.assertEqual(selector_unicost.seed, selector_diverse.seed)
```
Cool test! thank you!!

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,775 @@
+        best_selected_features = best_selector.transform(data)
+
+        # Verify the selected indices
+        self.assertListEqual(list(best_selector.transform(data).columns), ['item2', 'item3', 'item5'])
```
did you mean to use "best_selected_features " from line 74 here?

zohrehraziei replied:
Yes, I edited it.

One other possible test could be to conduct four different tests, each with the same number of features and cost metrics, but with a different optimization method. For example, we could create a simulation for each optimization method, with 10 different seeds, and then find the mean and variance of the different optimization techniques. By comparing the results, we could suggest which method is most robust against randomness. Implementing this improvement might take some additional time, but if you're interested in the idea, I could work on it and provide the results.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,775 @@
+        labels_infeasible = pd.DataFrame({"item1": [0, 0, 0, 0, 0], "item2": [0, 0, 0, 0, 0], "item3": [0, 0, 0, 0, 0],
+                                          "item4": [0, 0, 0, 0, 0], "item5": [0, 0, 0, 0, 0]})
+
+        method_infeasible = SelectionMethod.TextBased(num_features=2,
```
learnign questions; when the labels is infeasible, what do we actually return? Is the final selection is just empty? or sth like that?

zohrehraziei replied:
When the labels are infeasible, the selector will return an empty dataframe as the final selection because no feature can be selected that covers all labels.

skadio replied:
Is this captured in the pydoc? We should add a comment about infeasible behavior

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,775 @@
+        # Verify that the features selected
+        self.assertListEqual(list(selected_features2.columns), ['item2', 'item6'])
+
+    ################################################
```
thanks for strengthening the invalid tests side!

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,775 @@
+
+    # Check Random with unicost and diverse (the same features should select)
+    # Check solution for large trails with the same seed (the same features should select)
+    def test_text_based_random_num_feature(self):
```
The only other test I would recommend is "test_usage_example" which can be the first test in this file that is identical to the usage example to ensure that our usage example works. Add that tests, see if the example works :D and tweak if necessary

zohrehraziei replied:
test_usage_example is added to the test_text.py. Also I edited README.md accordingly.

zohrehraziei commented:
Hi Serdar,

I would appreciate it if you review the responses to the remaining reviews. I think three major points need to be confirmed:
1- Comment regarding the _get_diversity_cost() method
2- Comment regarding the _get_text_embeddings() method
3- Finalize the pydoc in the selector.py

Best,
Zohreh

skadio commented on README.md:
```diff
@@ -102,6 +108,42 @@ stats_df = calculate_statistics(score_df, selected_df)
+                     "article_4": ["article text here"]})
+
+# Labels to denote 0/1 coverage metadata for each article 
+# across three labels, e.g., sports, international, entertainment    
```
Here we are saying 3 labels but the example is changed to 4 labels. Can we please change this example so that the number of articles and labels are of different size, and not 4 vs. 4. This might unintentionally suggest a square matrix here. That's why I went with 4 articles and 3 labels. If the tests does not work that way, try 5 artciles, 4 labels ?

zohrehraziei replied:
That's a good point! I changed it with 5 articles, 4 labels.

skadio commented on tests/test_text.py:
```diff
@@ -0,0 +1,811 @@
+        # Verify the selected indices
+        self.assertListEqual(list(selected_features.columns), ['article_1', 'article_3'])
+
+    def test_categorical_data(self):
```
@zohrehraziei The only thing that remains is to check that, we can reproduce the examples/tables from the papers. The original problem definition includes "categories" which we flattened out in this library. 

What we need to do is to take a simple example df, that "has" categories, call the _process_category_data() function from the original code and assert that things work.

@bkleyn  can you provide a coverage labels df in the old style to test here? Something small, only to ensure that, given the old process() function, we can run things in this Library. 

In theory, if we run a df with categories (using process) and the same df given as flattened input (as is here) we should get the SAME output. This can be checked here  --and with that we can merge this.

zohrehraziei replied:
I have added the test as per the requirements for README.me. I have also ensured that all tests, including other selection methods, are run to ensure comprehensive testing of the code. 

For df with categories, I have created a simple input df with three articles and three categories. I then called the _process_category_data function with this df, passing in the categories and feature column. I have also added assertions to ensure that the resulting matrix and features have the expected shapes and values.
@bkleyn I kindly request your review of this test and confirmation if it follows the data format used in the paper.

skadio reviewed (approved):
@zohrehraziei things are looking great and we are almost there to merge it. 

Only two things remain: 
1) See my comment about the usage example in Tests and in Readme. It would be best if the input labels is not a square matrix. 
2) The last thing we need to ensure is that the inputs/problems from the paper works. I left a detailed commet in Tests about this.

With these two changes, we are good to merge :) Yay!!

zohrehraziei commented:
I have added the test as per the requirements for README.me. I have also ensured that all tests, including other selection methods, are run to ensure comprehensive testing of the code.

For df with categories, I have created a simple input df with three articles and three categories. I then called the _process_category_data function with this df, passing in the categories and feature column. I have also added assertions to ensure that the resulting matrix and features have the expected shapes and values.
@bkleyn I kindly request your review of this test and confirmation if it follows the data format used in the paper.
